codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -510,22 +511,26 @@ class JsonParserGenerator(
                                 """,
                                 *codegenScope,
                             )
-                            withBlock("variant = match key.to_unescaped()?.as_ref() {", "};") {
+                            withBlock("match key.to_unescaped()?.as_ref() {", "};") {
                                 for (member in shape.members()) {
                                     val variantName = symbolProvider.toMemberName(member)
                                     rustBlock("${jsonName(member).dq()} =>") {
                                         if (member.isTargetUnit()) {
                                             rustTemplate(
                                                 """
                                                 #{skip_value}(tokens)?;
-                                                Some(#{Union}::$variantName)
+                                                variant = Some(#{Union}::$variantName)
                                                 """,
                                                 "Union" to returnSymbolToParse.symbol, *codegenScope,
                                             )
                                         } else {
-                                            withBlock("Some(#T::$variantName(", "))", returnSymbolToParse.symbol) {
+                                            withBlock("variant = Some(#T::$variantName(", "));", returnSymbolToParse.symbol) {
                                                 deserializeMember(member)
-                                                unwrapOrDefaultOrError(member)
+                                                if (checkValueSet) {
+                                                    rust(""".ok_or_else(|| aws_smithy_json::deserialize::error::DeserializeError::custom("no variant was set in union"))?""")
```

Comment:
This can be checked in `unwrapOrDefaultOrError`. In `unwrapOrDefaultOrError` the error is returned as "this member cannot be null", but here we're checking that "one (any) member must be set". I can move the check inside `unwrapOrDefaultOrError`